File: AEDS-2/TP02/TP02Q02 - Registro em C/Principal.c
/*
   ==UserScript==
 @name         TP02Q02 - Registro em C
 @namespace    https://github.com/Ddavidi/PUC-MINAS
 @description  VERDE PUC MINAS - TP02Q02 - Registro em C
 @author       @ddavidi_
   ==/UserScript==
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SHOWS 1370
#define MAX_LINE 1024
#define MAX_CAST 30
#define MAX_NAME 100

typedef struct {
    char show_id[20];
    char type[50];
    char title[200];
    char director[100];
    char cast[300];
    char country[100];
    char date_added[50];
    int release_year;
    char rating[20];
    char duration[20];
    char listed_in[200];
} Show;

void trim_newline(char *str) {
    int i = 0;
    while (str[i]) {
        if (str[i] == '\n') {
            str[i] = '\0';
            return;
        }
        i++;
    }
}

void copiarCampo(char *destino, const char *linha, int *i) {
    int j = 0;
    if (linha[*i] == '"') {
        (*i)++;
        while (linha[*i] && !(linha[*i] == '"' && linha[*i + 1] == ',')) {
            if (linha[*i] == '"' && linha[*i + 1] == '"') {
                // Ignorar as aspas duplas
                (*i) += 2;
            } else {
                destino[j++] = linha[*i];
                (*i)++;
            }
        }
        (*i) += 2; 
    } else {
        while (linha[*i] && linha[*i] != ',' && linha[*i] != '\n') {
            destino[j++] = linha[*i];
            (*i)++;
        }
        if (linha[*i] == ',') (*i)++; 
    }
    destino[j] = '\0';
}
const char* verificaCampo(const char* campo) {
    return (campo[0] == '\0') ? "NaN" : campo;
}

void ordenar_cast(char cast_ordenado[][MAX_NAME], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - 1 - i; j++) {
            if (strcmp(cast_ordenado[j], cast_ordenado[j + 1]) > 0) {
                char temp[MAX_NAME];
                strcpy(temp, cast_ordenado[j]);
                strcpy(cast_ordenado[j], cast_ordenado[j + 1]);
                strcpy(cast_ordenado[j + 1], temp);
            }
        }
    }
}

void imprimirShow(Show s) {
    // Processar e ordenar cast
    char cast_ordenado[MAX_CAST][MAX_NAME];
    int cast_count = 0;

    if (s.cast[0] != '\0') {
        int i = 0, j = 0, k = 0;
        while (s.cast[i]) {
            if (s.cast[i] == ',' && s.cast[i + 1] == ' ') {
                cast_ordenado[cast_count][j] = '\0';
                cast_count++;
                i += 2;
                j = 0;
            } else {
                cast_ordenado[cast_count][j++] = s.cast[i++];
            }
        }
        cast_ordenado[cast_count][j] = '\0';
        cast_count++;

        ordenar_cast(cast_ordenado, cast_count);
    }

    // Imprimir formatado
    printf("=> %s ## %s ## %s ## %s ## [",
           verificaCampo(s.show_id),
           verificaCampo(s.title),
           verificaCampo(s.type),
           verificaCampo(s.director));

    if (cast_count > 0) {
        for (int i = 0; i < cast_count; i++) {
            printf("%s", verificaCampo(cast_ordenado[i]));
            if (i < cast_count - 1) printf(", ");
        }
    } else {
        printf("NaN");
    }

    printf("] ## %s ## %s ## %d ## %s ## %s ## [%s] ##\n",
           verificaCampo(s.country),
           verificaCampo(s.date_added),
           s.release_year > 0 ? s.release_year : 0,
           verificaCampo(s.rating),
           verificaCampo(s.duration),
           verificaCampo(s.listed_in));
}

int main() {
    FILE *file = fopen("/tmp/disneyplus.csv", "r");
    if (file == NULL) {
        printf("Erro ao abrir o arquivo.\n");
        return 1;
    }

    Show shows[MAX_SHOWS];
    char linha[MAX_LINE];
    int count = 0;

    fgets(linha, MAX_LINE, file); // Pula cabeçalho

    while (fgets(linha, MAX_LINE, file) && count < MAX_SHOWS) {
        int i = 0;
        copiarCampo(shows[count].show_id, linha, &i);
        copiarCampo(shows[count].type, linha, &i);
        copiarCampo(shows[count].title, linha, &i);
        copiarCampo(shows[count].director, linha, &i);
        copiarCampo(shows[count].cast, linha, &i);
        copiarCampo(shows[count].country, linha, &i);
        copiarCampo(shows[count].date_added, linha, &i);

        char release_year_str[10];
        copiarCampo(release_year_str, linha, &i);
        shows[count].release_year = atoi(release_year_str);

        copiarCampo(shows[count].rating, linha, &i);
        copiarCampo(shows[count].duration, linha, &i);
        copiarCampo(shows[count].listed_in, linha, &i);

        count++;
    }

    fclose(file);
    char entrada[20];
    fgets(entrada, sizeof(entrada), stdin);
    trim_newline(entrada);

    while (strcmp(entrada, "FIM") != 0) {
        int encontrado = 0;
        for (int i = 0; i < count; i++) {
            if (strcmp(shows[i].show_id, entrada) == 0) {
                imprimirShow(shows[i]);
                encontrado = 1;
                break;
            }
        }

        fgets(entrada, sizeof(entrada), stdin);
        trim_newline(entrada);
    }

    return 0;
}

File: AEDS-2/TP02/TP02Q15 - Seleção Parcial em Java/Principal.java
/*
   ==UserScript==
 @name         TP02Q15 - Seleção Parcial em Java
 @namespace    https://github.com/Ddavidi/PUC-MINAS
 @description  VERDE PUC MINAS - TP02Q15 - Seleção Parcial em Java
 @author       @ddavidi_
   ==/UserScript==
*/

import java.io.File;
import java.util.Arrays;
import java.util.Scanner;

public class Principal {
    public static void main(String[] args) {
        Show[] lista = Show.lerArquivo();
        Scanner sc = new Scanner(System.in);
        String entrada = sc.nextLine();
        Show[] arrayShows = new Show[1370];
        int i = 0;

        while (!entrada.equals("FIM")) {
            try {
                int index = Integer.parseInt(entrada.substring(1));
                if (index >= 0 && index < lista.length && lista[index] != null) {
                    try {
                        arrayShows[i++] = lista[index - 1].clone();
                    } catch (CloneNotSupportedException e) {
                        System.out.println("Erro ao clonar o objeto Show: " + e.getMessage());
                    }
                } else {
                    System.out.println("Índice inválido ou show não encontrado.");
                }
            } catch (NumberFormatException e) {
                System.out.println("Erro: entrada inválida. Certifique-se de usar o formato correto (Xn).");
            } catch (ArrayIndexOutOfBoundsException | NullPointerException e) {
                System.out.println("Erro inesperado: " + e.getMessage());
            }
            entrada = sc.nextLine();
        }

        Show.insertionSort(arrayShows, i);
        for (int j = 0; j < 10; j++) {
            if (arrayShows[j] != null) {
                arrayShows[j].imprimir();
            }
        }

        sc.close();
    }
}

class Show implements Cloneable {
    private String show_id;
    private String type;
    private String title;
    private String director;
    private String[] cast;
    private String country;
    private Date date_added;
    private int release_year;
    private String rating;
    private String duration;
    private String[] listed_in;
    private String description;

    public Show() {
        this.show_id = "NaN";
        this.type = "NaN";
        this.title = "NaN";
        this.director = "NaN";
        this.cast = new String[1];
        this.country = "NaN";
        this.date_added = new Date();
        this.release_year = 1900;
        this.rating = "NaN";
        this.duration = "NaN";
        this.listed_in = new String[1];
        this.description = "NaN";
    }

    public Show(String show_id, String type, String title, String director, String[] cast, String country,
                Date date_added, int release_year, String rating, String duration, String[] listed_in, String description) {
        this.show_id = show_id;
        this.type = type;
        this.title = title;
        this.director = director;
        this.cast = Arrays.copyOf(cast, cast.length);
        this.country = country;
        this.date_added = date_added;
        this.release_year = release_year;
        this.rating = rating;
        this.duration = duration;
        this.listed_in = Arrays.copyOf(listed_in, listed_in.length);
        this.description = description;
    }
    
    // Getters
    public String getShow_id() { return show_id; }
    public String getType() { return type; }
    public String getTitle() { return title; }
    public String getDirector() { return director; }
    public String[] getCast() { return cast; }
    public String getCountry() { return country; }
    public Date getDate_added() { return date_added; }
    public int getRelease_year() { return release_year; }
    public String getRating() { return rating; }
    public String getDuration() { return duration; }
    public String[] getListed_in() { return listed_in; }
    public String getDescription() { return description; }

    // Setters
    public void setShow_id(String show_id) { this.show_id = show_id; }
    public void setType(String type) { this.type = type; }
    public void setTitle(String title) { this.title = title; }
    public void setDirector(String director) { this.director = director; }
    public void setCast(String[] cast) { this.cast = Arrays.copyOf(cast, cast.length); }
    public void setCountry(String country) { this.country = country; }
    public void setDate_added(Date date_added) { this.date_added = date_added; }
    public void setRelease_year(int release_year) { this.release_year = release_year; }
    public void setRating(String rating) { this.rating = rating; }
    public void setDuration(String duration) { this.duration = duration; }
    public void setListed_in(String[] listed_in) { this.listed_in = Arrays.copyOf(listed_in, listed_in.length); }
    public void setDescription(String description) { this.description = description; }

    // Clone
    @Override
    public Show clone() throws CloneNotSupportedException {
        return (Show) super.clone();
    }

    public static Show[] lerArquivo() {
        Show[] lista = new Show[1370];
        int index = 0;
        File arquivo = new File("/tmp/disneyplus.csv");

        try (Scanner sc = new Scanner(arquivo)) {
            sc.nextLine(); // pula o cabeçalho do arquivo CSV
            while (sc.hasNextLine()) {
                String linha = sc.nextLine();
                lista[index] = new Show();
                lista[index++].ler(linha);
            }
        } catch (Exception e) {
            System.out.println("Erro ao ler o arquivo: " + e.getMessage());
        }
        return lista;
    }

    public void ler(String linha) {
        try {
            String[] partes = linha.split(",(?=([^\"]*\"[^\"]*\")*[^\"]*$)", -1);
            this.show_id = partes[0].trim().isEmpty() ? "NaN" : partes[0].trim();
            this.type = partes[1].trim().isEmpty() ? "NaN" : partes[1].replace("\"", "").trim();
            this.title = partes[2].trim().isEmpty() ? "NaN" : partes[2].replace("\"", "").trim();
            this.director = partes[3].trim().isEmpty() ? "NaN" : partes[3].replace("\"", "").trim();
            this.cast = partes[4].trim().isEmpty() ? new String[]{"NaN"} : separarString(partes[4]);
            ordenarCast(this.cast);
            this.country = partes[5].trim().isEmpty() ? "NaN" : partes[5].replace("\"", "").trim();
            this.date_added = Date.parseData(partes[6].trim());
            this.release_year = partes[7].trim().isEmpty() ? 1900 : Integer.parseInt(partes[7].trim());
            this.rating = partes[8].trim().isEmpty() ? "NaN" : partes[8].trim();
            this.duration = partes[9].trim().isEmpty() ? "NaN" : partes[9].trim();
            this.listed_in = partes[10].trim().isEmpty() ? new String[]{"NaN"} : separarString(partes[10]);
        } catch (NumberFormatException | ArrayIndexOutOfBoundsException | NullPointerException e) {
            System.out.println("Erro ao processar linha: " + linha + " - " + e.getMessage());
        }
    }

    private String[] separarString(String str) {
        String[] partes = str.replace("\"", "").split(", ");
        return partes;
    }

    public void imprimir() {
        System.out.printf("=> %s ## %s ## %s ## %s ## [", show_id, title, type, director);

        for (int i = 0; i < cast.length; i++) {
            System.out.printf("%s", cast[i]);
            if (i < cast.length - 1) System.out.printf(", ");
        }

        System.out.printf("] ## %s ## %s ## %d ## %s ## %s ## [", country, date_added.toString(), release_year, rating, duration);

        for (int i = 0; i < listed_in.length; i++) {
            System.out.printf("%s", listed_in[i]);
            if (i < listed_in.length - 1) System.out.printf(", ");
        }

        System.out.printf("] ##\n");
    }

    public static void swap(String [] cast, int i, int j) {
        String aux = cast[i];
        cast[i] = cast[j];
        cast[j] = aux;
    }

    public static void ordenarCast(String[] cast) {
        for(int i = 0; i < cast.length; i++){
            int menor  = i;
            for(int j = i + 1; j < cast.length; j++){
                if(cast[menor].compareTo(cast[j]) > 0){
                    menor = j;
                }
            }
            if(menor != i){
                swap(cast, i, menor);
            }
        }
    }

        public static void insertionSort(Show[] array, int tam) {
        for (int i = 1; i < tam; i++) {
            Show temp = array[i];
            int j = i - 1;

            while (j >= 0 && compare(temp, array[j]) < 0) {
                array[j + 1] = array[j];
                j--;
            }

            array[j + 1] = temp;
        }
    }

    private static int compare(Show s1, Show s2) {
        return s1.title.compareTo(s2.title);
    }
}

class Date {
    private String month;
    private int day;
    private int year;

    public Date() {
        this("March", 1, 1900);
    }

    public Date(String month, int day, int year) {
        this.month = month;
        this.day = day;
        this.year = year;
    }

    public void setMonth(String month) { this.month = month; }
    public void setDay(int day) { this.day = day; }
    public void setYear(int year) { this.year = year; }

    public String getMonth() { return month; }
    public int getDay() { return day; }
    public int getYear() { return year; }

    public static Date parseData(String dataTexto) {
        if (dataTexto == null || dataTexto.trim().isEmpty() || dataTexto.equals("NaN")) {
            return new Date();
        }

        try {
            // Remove aspas, se existirem, e divide a string
            dataTexto = dataTexto.replace("\"", "").trim();
            String[] partes = dataTexto.split(" ");

            if (partes.length == 3) {
                String month = partes[0].trim();
                int day = Integer.parseInt(partes[1].replace(",", "").trim());
                int year = Integer.parseInt(partes[2].trim());
                return new Date(month, day, year);
            } else {
                throw new IllegalArgumentException("Formato de data inválido: " + dataTexto);
            }
        } catch (NumberFormatException e) {
            System.out.println("Erro ao processar data (NumberFormatException): " + dataTexto);
            return new Date(); // Retorna uma data padrão em caso de erro
        } catch (IllegalArgumentException e) {
            System.out.println("Erro ao processar data (IllegalArgumentException): " + dataTexto);
            return new Date(); // Retorna uma data padrão em caso de erro
        }
    }

    @Override
    public String toString() {
        return String.format("%s %d, %04d", month, day, year);
    }
}